plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php

```diff
@@ -297,29 +297,19 @@ private static function order_received( $order_id = 0 ) {
 			return;
 		}
 
+		$template_args     = array( 'order' => $order );
 		$order_customer_id = $order->get_customer_id();
 
-		// For non-guest orders, require the user to be logged in before showing this page.
-		if ( $order_customer_id && get_current_user_id() !== $order_customer_id ) {
-			wc_print_notice( esc_html__( 'Please log in to your account to view this order.', 'woocommerce' ), 'notice' );
-			woocommerce_login_form( array( 'redirect' => $order->get_checkout_order_received_url() ) );
-			return;
-		}
-
-		// For guest orders, request they verify their email address (unless we can identify them via the active user session).
-		if ( self::guest_should_verify_email( $order, 'order-received' ) ) {
-			wc_get_template(
-				'checkout/form-verify-email.php',
-				array(
-					'failed_submission' => ! empty( $_POST['email'] ), // phpcs:ignore WordPress.Security.NonceVerification.Missing
-					'verify_url'        => $order->get_checkout_order_received_url(),
-				)
-			);
-			return;
+		if ( ! $order_customer_id && self::guest_should_verify_email( $order, 'order-received' ) ) {
+			// For guest orders, request they verify their email address (unless we can identify them via the active user session).
+			$template_args['needs_verification'] = 'email';
+		} elseif ( $order_customer_id && get_current_user_id() !== $order_customer_id ) {
+			// For non-guest orders, require the user to be logged in before showing this page.
+			$template_args['needs_verification'] = 'login';
```

Comment:
(2/3) On the basis that these conditions may not be satisfied and therefore `$template_args['needs_verification']` may not be set.